File: services/api/internal/professional/domain/professional.go
package domain

import (
	"errors"
	"time"

	"github.com/google/uuid"
)

// RegistrationType defines the professional registration body.
type RegistrationType string

const (
	RegistrationCRN   RegistrationType = "CRN"
	RegistrationCRM   RegistrationType = "CRM"
	RegistrationOther RegistrationType = "other"
)

// ServiceMode defines how a professional delivers care.
type ServiceMode string

const (
	ServiceModeOnsite    ServiceMode = "onsite"
	ServiceModeOnline    ServiceMode = "online"
	ServiceModeHomeVisit ServiceMode = "home_visit"
)

// Sentinel errors.
var (
	ErrNotFound              = errors.New("professional: not found")
	ErrDuplicateRegistration = errors.New("professional: duplicate registration")
	ErrDuplicateUser         = errors.New("professional: user already registered in tenant")
	ErrAddressNotFound       = errors.New("professional: address not found")
	ErrServiceModeNotFound   = errors.New("professional: service mode not found")
	ErrEntitlementExceeded   = errors.New("professional: entitlement limit exceeded")
)

// Professional represents a healthcare professional within a tenant.
type Professional struct {
	ID                 uuid.UUID
	TenantID           uuid.UUID
	UserID             uuid.UUID
	FullName           string
	RegistrationType   RegistrationType
	RegistrationNumber string
	RegistrationState  string
	Specialty          string
	Bio                string
	Phone              string
	AvatarURL          string
	Active             bool
	CreatedAt          time.Time
	UpdatedAt          time.Time
}

// Validate checks business rules for a Professional.
func (p *Professional) Validate() error {
	if p.FullName == "" {
		return errors.New("professional: full_name is required")
	}
	if p.RegistrationType == "" {
		return errors.New("professional: registration_type is required")
	}
	if p.RegistrationNumber == "" {
		return errors.New("professional: registration_number is required")
	}
	if (p.RegistrationType == RegistrationCRN || p.RegistrationType == RegistrationCRM) && p.RegistrationState == "" {
		return errors.New("professional: registration_state is required for CRN/CRM")
	}
	return nil
}

// Address represents a physical location where a professional provides care.
type Address struct {
	ID             uuid.UUID
	TenantID       uuid.UUID
	ProfessionalID uuid.UUID
	Label          string
	Street         string
	Number         string
	Complement     string
	Neighborhood   string
	City           string
	State          string
	ZipCode        string
	Country        string
	Latitude       *float64
	Longitude      *float64
	Phone          string
	Notes          string
	Active         bool
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

// Validate checks business rules for an Address.
func (a *Address) Validate() error {
	if a.Label == "" {
		return errors.New("address: label is required")
	}
	if a.Street == "" {
		return errors.New("address: street is required")
	}
	if a.City == "" {
		return errors.New("address: city is required")
	}
	if a.State == "" {
		return errors.New("address: state is required")
	}
	if a.ZipCode == "" {
		return errors.New("address: zip_code is required")
	}
	return nil
}

// ProfessionalServiceMode represents a service mode for a professional.
type ProfessionalServiceMode struct {
	ID             uuid.UUID
	TenantID       uuid.UUID
	ProfessionalID uuid.UUID
	Mode           ServiceMode
	AddressID      *uuid.UUID
	DurationMin    int
	Active         bool
	CreatedAt      time.Time
	UpdatedAt      time.Time
}

File: apps/web-professional/src/lib/schemas/professional.ts
import { z } from "zod"

export const profileSchema = z.object({
  full_name: z.string().min(2, "Nome completo é obrigatório"),
  registration_type: z.string().min(1, "Tipo de registro é obrigatório"),
  registration_number: z.string().min(1, "Número de registro é obrigatório"),
  registration_state: z.string().optional().default(""),
  specialty: z.string().optional().default(""),
  bio: z.string().optional().default(""),
  phone: z.string().optional().default(""),
})

export type ProfileFormValues = z.infer<typeof profileSchema>

export const registrationTypes = [
  { value: "CRN", label: "CRN (Nutricionista)" },
  { value: "CRM", label: "CRM (Médico)" },
  { value: "other", label: "Outro" },
] as const

export const brazilianStates = [
  "AC","AL","AP","AM","BA","CE","DF","ES","GO","MA","MT","MS","MG",
  "PA","PB","PR","PE","PI","RJ","RN","RS","RO","RR","SC","SP","SE","TO",
] as const

export const addressSchema = z.object({
  label: z.string().min(1, "Nome do endereço é obrigatório"),
  street: z.string().min(1, "Rua é obrigatória"),
  number: z.string().optional().default(""),
  complement: z.string().optional().default(""),
  neighborhood: z.string().optional().default(""),
  city: z.string().min(1, "Cidade é obrigatória"),
  state: z.string().min(2, "Estado é obrigatório"),
  zip_code: z.string().min(8, "CEP é obrigatório"),
  country: z.string().optional().default("BR"),
  phone: z.string().optional().default(""),
  notes: z.string().optional().default(""),
})

export type AddressFormValues = z.infer<typeof addressSchema>

export const serviceModeSchema = z.object({
  mode: z.enum(["onsite", "online", "home_visit"]),
  address_id: z.string().optional(),
  duration_min: z.coerce.number().min(1, "Duração mínima é 1 minuto"),
})

export type ServiceModeFormValues = z.infer<typeof serviceModeSchema>

export const serviceModeLabels: Record<string, string> = {
  onsite: "Presencial",
  online: "Online",
  home_visit: "Domiciliar",
}
